### PR TITLE
[WIP] Add method to modify calculate fee's coefficient

### LINKED ISSUE #2004: change the longest chain switch length
https://github.com/AElfProject/AElf/blob/572e1b5497c787c3e87a1f8c455f740ec714aef0/src/AElf.Kernel.Core/Blockchain/Domain/IChainManager.cs#L153-L158

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a `TODO` comment in 572e1b5497c787c3e87a1f8c455f740ec714aef0 when #1992 was merged. cc @AElfProject.

### PR DESCRIPTION
**Pull request by llypb**
1. add 5 kinds(4 resource token + tx fee) of strategy with different algorithm. 
2. add interface to modify the coefficient of algorithm using to calculate fee.

Discussed with team, the isssue does not need to be implemented, so close #2004

### CODE REVIEW

File: src/AElf.Kernel.TransactionPool/Application/CalculateFeeService.cs [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+            IBlockchainService blockchainService,
+            IChainBlockLinkService chainBlockLinkService)
+        {
+            _tokenStTokenContractReaderFactory = tokenStTokenContractReaderFactory;
+            _blockchainService = blockchainService;
+            _chainBlockLinkService = chainBlockLinkService;
+            Logger = new NullLogger<CalculateAlgorithm>();
+        }
+
+        private readonly Dictionary<int, ICalculateWay> _defaultPieceWiseFunc = new Dictionary<int, ICalculateWay>();
```
Comment (llypb) [inline src/AElf.Kernel.TransactionPool/Application/CalculateFeeService.cs]: remove

File: src/AElf.Kernel.TransactionPool/Application/CalculateFeeService.cs [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+        private Dictionary<int, ICalculateWay> _pieceWiseFuncCache = null;
```
Comment (llypb) [inline src/AElf.Kernel.TransactionPool/Application/CalculateFeeService.cs]: add provider

File: src/AElf.Kernel.TransactionPool/Application/CalculateFeeService.cs [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+        {
+            var funcDic = new Dictionary<int, ICalculateWay>();
+            foreach (var func in allFunc)
+            {
+                funcDic[func.PieceKey] = func;
+            }
+
+            _forkCache[blockIndex] = funcDic;
+        }
+        private void AddPieceFunction(int pieceKey, IDictionary<int, ICalculateWay> pieceWiseFunc,
```
Comment (llypb) [inline src/AElf.Kernel.TransactionPool/Application/CalculateFeeService.cs]: remove

File: src/AElf.Kernel.TransactionPool/Application/CalculateFeeService.cs [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+            else
+            {
+                parameters = await tokenStub.GetCalculateFeeCoefficientOfContract.CallAsync(new SInt32Value
+                    {Value = CalculateAlgorithmContext.CalculateFeeTypeEnum});
+            }
+
+            if (parameters == null)
+            {
+                Logger.LogWarning("does not find parameter from contract, initialize from default ");
+                _pieceWiseFuncCache = _defaultPieceWiseFunc.ToDictionary(x => x.Key, x => x.Value);
```
Comment (llypb) [inline src/AElf.Kernel.TransactionPool/Application/CalculateFeeService.cs]: remove

File: src/AElf.Kernel.TransactionPool/Application/CalculateFeeService.cs [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+                {
+                    Logger.LogWarning($"could not find mapped function type {func.FunctionType}");
+                    continue;
+                }
+
+                calWayDic[func.PieceKey] = newCalculateWay;
+                calWayDic[func.PieceKey].InitParameter(func.CoefficientDic);
+            }
+
+            _forkCache[CalculateAlgorithmContext.BlockIndex] = calWayDic;
```
Comment (llypb) [inline src/AElf.Kernel.TransactionPool/Application/CalculateFeeService.cs]: remove

File: src/AElf.Kernel.TransactionPool/Application/CalculateWay.cs [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+using System.Collections.Generic;
+using AElf.Contracts.MultiToken;
+using AElf.Kernel.SmartContract.Application;
+using AElf.Sdk.CSharp;
+
+namespace AElf.Kernel.TransactionPool.Application
+{
+    #region ICalculateWay implemention   
+
+    public class LnCalculateWay : ICalculateWay
```
Comment (llypb) [inline src/AElf.Kernel.TransactionPool/Application/CalculateWay.cs]: remove unused clay

File: contract/AElf.Contracts.MultiToken/TokenContract_Fees.cs [hunk 7]
```diff
@@ -434,6 +434,21 @@ public override Empty SetFeeReceiver(Address input)
         }
 
         public override Empty SetTransactionSizeUnitPrice(SInt64Value input)
+        {
+            AssertIsAuthorized();
+            Context.Fire(new TransactionSizeFeeUnitPriceUpdated
+            {
+                UnitPrice = input.Value
+            });
```
Comment (rosona) [inline contract/AElf.Contracts.MultiToken/TokenContract_Fees.cs]: Context.Fire after set value.

File: contract/AElf.Contracts.MultiToken/TokenContract_Fee_Calculate_Coefficient.cs [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+            });
+            return new Empty();
+        }
+
+        public override Empty UpdateCoefficientFromSender(CoefficientFromSender coeInput)
+        {
+            if (coeInput == null)
+                return new Empty();
+            AssertIsAuthorized();
+            var dataInDb = State.CalculateCoefficientOfSender.Value;
```
Comment (rosona) [inline contract/AElf.Contracts.MultiToken/TokenContract_Fee_Calculate_Coefficient.cs]: rename dateInDb, theOne.

File: src/AElf.Kernel.TransactionPool/Application/CalculateFeeService.cs [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+
+    interface ICalculateFunctionProvider : ITransientDependency
+    {
+        Dictionary<int, ICalculateWay> PieceWiseFuncCache { get; set; }
+        ConcurrentDictionary<BlockIndex, Dictionary<int, ICalculateWay>> ForkCache { get; set; }
+    }
+    class CalculateFunctionProvider : ICalculateFunctionProvider
+    {
+        public Dictionary<int, ICalculateWay> PieceWiseFuncCache { get; set; }
+        public Dictionary<int, ICalculateWay> DefaultWiseFuncCache { get; set; } = new Dictionary<int, ICalculateWay>();
```
Comment (llypb) [inline src/AElf.Kernel.TransactionPool/Application/CalculateFeeService.cs]: remove

**Pull request merged**